File: func_log/zadania1_5/zadanie2.py
#нефункциональный

def collatz_imperative(n):
    sequence = [n]          # изменяемое состояние
    while n != 1:
        if n % 2 == 0:
            n = n // 2       # мутация переменной
        else:
            n = n * 3 + 1
        sequence.append(n)   # мутация списка
    return sequence


if __name__ == "__main__":
    n = int(input("Введите положительное целое число n: "))
    print(collatz_imperative(n))

    
#функциональный

def collatz_functional(n):
    def step(x):
        # чистая функция: результат зависит только от аргумента
        return x // 2 if x % 2 == 0 else x * 3 + 1

    def go(x, acc):
        # рекурсия вместо цикла, иммутабельное расширение списка
        if x == 1:
            return acc
        next_x = step(x)
        return go(next_x, acc + [next_x])

    return go(n, [n])


if __name__ == "__main__":
    n = int(input("Введите положительное целое число n: "))
    print(collatz_functional(n))

File: func_log/zadania1_5/zadanie2.js
//нефункциональный 

function collatzImperative(n) {
    const sequence = [n];                       // изменяемое состояние
    while (n !== 1) {
        if (n % 2 === 0) {
            n = n / 2;                          // мутация переменной
        } else {
            n = n * 3 + 1;
        }
        sequence.push(n);                       // мутация массива
    }
    return sequence;
}

// Пример использования
const n = 5;
console.log(collatzImperative(n));              // [5, 16, 8, 4, 2, 1]

// функциональный
const collatzFunctional = n => {
    // чистая функция: результат зависит только от аргумента
    const step = x => (x % 2 === 0 ? x / 2 : x * 3 + 1);

    // рекурсия вместо цикла, иммутабельное расширение массива
    const go = (x, acc) => {
        if (x === 1) return acc;
        const nextX = step(x);
        return go(nextX, [...acc, nextX]);
    };

    return go(n, [n]);
};

// Пример использования
const n = 5;
console.log(collatzFunctional(n));              // [5, 16, 8, 4, 2, 1]
